test_utils/lib/src/replay_http_client.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:collection/collection.dart';
+import 'package:http/http.dart';
+
+import 'model.dart';
+import 'test_http_client.dart';
+
+class ReplayHttpClient extends TestHttpClient {
+  List<(RecordedRequest, RecordedResponse)> _requestResponse = [];
+
+  ReplayHttpClient();
+
+  @override
+  Future<void> startTest(String packageName, String test) async {
+    _requestResponse = RecordedSession.fromJson(
+      jsonDecode(
+        await File(
+          await TestHttpClient.recordPath(packageName, test),
+        ).readAsString(),
+      ),
+    ).requestResponse;
+  }
+
+  @override
+  Future<void> endTest() async {
+    if (_requestResponse.isNotEmpty) {
+      throw StateError('test completed without sending all expected requests');
+    }
+  }
+
+  void _matchRequest(
+    RecordedRequest recordedRequest,
+    BaseRequest request,
+    List<int> body,
+  ) {
+    if (recordedRequest.url != request.url) {
+      throw StateError(
+        'recorded request URL ${recordedRequest.url} '
+        'does not match ${request.url}',
+      );
+    }
+
+    if (recordedRequest.method != request.method) {
+      throw StateError(
+        'recorded request method ${recordedRequest.method} '
+        'does not match ${request.method}',
+      );
+    }
+
+    // Don't consider "x-goog-api-client" because it is non-semantic and
+    // varies based on the current Dart version.
+    final recordedHeaders = Map<String, String>.from(recordedRequest.headers)
+      ..remove('x-goog-api-client');
+    final requestHeaders = Map<String, String>.from(request.headers)
+      ..remove('x-goog-api-client');
```

Comment:
[nit] Map literals have become more idiomatic than `Map.from`.


```suggestion
    final recordedHeaders = {...recordedRequest.headers}
      ..remove('x-goog-api-client');
    final requestHeaders = {...request.headers}
      ..remove('x-goog-api-client');
```